src/fiber/graph.rs

```diff
@@ -826,6 +859,60 @@ pub enum PaymentSessionStatus {
     Failed,
 }
 
+#[serde_as]
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct SessionRouteNode {
+    pub pubkey: Pubkey,
+    pub amount: u128,
+    #[serde_as(as = "EntityHex")]
+    pub channel_outpoint: OutPoint,
+}
+
+// The router is a list of nodes that the payment will go through.
+// We store in the payment session and then will use it to track the payment history.
+// The router is a list of nodes that the payment will go through.
+// For example:
+//    A(amount, channel) -> B -> C -> D means A will send `amount` with `channel` to B.
+#[derive(Clone, Debug, Serialize, Deserialize, Default)]
+pub struct SessionRoute {
+    pub nodes: Vec<SessionRouteNode>,
+}
+
+impl SessionRoute {
+    // Create a new route from the source to the target with the given payment hops.
+    // The payment hops are the hops that the payment will go through.
+    // for a payment route A -> B -> C -> D
+    // the `payment_hops` is [B, C, D], which is a convinent way for onion routing.
+    // here we need to create a session route with source, which is A -> B -> C -> D
+    pub fn new(source: Pubkey, target: Pubkey, payment_hops: &Vec<PaymentHopData>) -> Self {
+        let mut router = Self::default();
+        let mut current = source;
+        for hop in payment_hops {
+            if let Some(key) = hop.next_hop {
+                router.add_node(
+                    current,
+                    hop.channel_outpoint
+                        .clone()
+                        .expect("expect channel outpoint"),
+                    hop.amount,
+                );
+                current = key;
+            }
+        }
+        assert_eq!(current, target);
+        router.add_node(target, OutPoint::default(), 0);
+        router
+    }
+
+    fn add_node(&mut self, pubkey: Pubkey, channel_outpoint: OutPoint, amount: u128) {
+        self.nodes.push(SessionRouteNode {
+            pubkey,
+            channel_outpoint,
+            amount,
+        });
+    }
```

Comment:
suggest to use zip to simplify code and remove unnecessary vec allocation code, btw, the type of payment_hops can be simplified to slice:

```suggestion
    pub fn new(source: Pubkey, target: Pubkey, payment_hops: &[PaymentHopData]) -> Self {
        let nodes = std::iter::once(source)
            .chain(payment_hops.iter().map(|hop| hop.next_hop.clone().unwrap_or(target)))
            .zip(payment_hops)
            .map(|(pubkey, hop)| SessionRouteNode {
                pubkey,
                channel_outpoint: hop.channel_outpoint.clone().unwrap_or_default(),
                amount: hop.amount,
            })
            .collect();

        Self { nodes }
    }
```